Colab-TextGen-GPU.ipynb

```diff
@@ -0,0 +1,144 @@
+        "  is_cuda118 = '+cu118' in torver  # 2.1.0+cu118\n",
+        "  is_cuda117 = '+cu117' in torver  # 2.0.1+cu117\n",
+        "\n",
+        "  textgen_requirements = open('requirements.txt').read().splitlines()\n",
+        "  if is_cuda117:\n",
+        "      textgen_requirements = [req.replace('+cu121', '+cu117').replace('+cu122', '+cu117').replace('torch2.1', 'torch2.0') for req in textgen_requirements]\n",
+        "  elif is_cuda118:\n",
+        "      textgen_requirements = [req.replace('+cu121', '+cu118').replace('+cu122', '+cu118') for req in textgen_requirements]\n",
+        "  with open('temp_requirements.txt', 'w') as file:\n",
+        "      file.write('\\n'.join(textgen_requirements))\n",
+        "\n",
+        "  !pip install -r extensions/openai/requirements.txt --upgrade\n",
+        "  !pip install -r temp_requirements.txt --upgrade\n",
+        "\n",
+        "  print(\"\\033[1;32;1m\\n --> If you see a warning about \\\"previously imported packages\\\", just ignore it.\\033[0;37;0m\")\n",
+        "  print(\"\\033[1;32;1m\\n --> There is no need to restart the runtime.\\n\\033[0;37;0m\")\n",
+        "\n",
+        "  try:\n",
+        "    import flash_attn\n",
+        "  except:\n",
+        "    !pip uninstall -y flash_attn\n",
+        "\n",
+        "# Parameters\n",
+        "model_url = \"https://huggingface.co/TheBloke/MythoMax-L2-13B-GPTQ\" #@param {type:\"string\"}\n",
+        "branch = \"gptq-4bit-32g-actorder_True\" #@param {type:\"string\"}\n",
+        "command_line_flags = \"--n-gpu-layers 128 --load-in-4bit --use_double_quant\" #@param {type:\"string\"}\n",
+        "api = False #@param {type:\"boolean\"}\n",
+        "\n",
+        "if api:\n",
+        "  for param in ['--api', '--public-api']:\n",
+        "    if param not in command_line_flags:\n",
+        "      command_line_flags += f\" {param}\"\n",
+        "\n",
+        "model_url = model_url.strip()\n",
+        "if model_url != \"\":\n",
+        "    if not model_url.startswith('http'):\n",
+        "        model_url = 'https://huggingface.co/' + model_url\n",
+        "\n",
+        "    # Download the model\n",
+        "    url_parts = model_url.strip('/').strip().split('/')\n",
+        "    output_folder = f\"{url_parts[-2]}_{url_parts[-1]}\"\n",
+        "    branch = branch.strip('\"\\' ')\n",
+        "    if branch.strip() not in ['', 'main']:\n",
+        "        output_folder += f\"_{branch}\"\n",
+        "        !python download-model.py {model_url} --branch {branch}\n",
+        "    else:\n",
+        "        !python download-model.py {model_url}\n",
+        "else:\n",
+        "    output_folder = \"\"\n",
+        "\n",
+        "# Start the web UI\n",
+        "cmd = f\"python server.py --share\"\n",
+        "if output_folder != \"\":\n",
+        "    cmd += f\" --model {output_folder}\"\n",
+        "cmd += f\" {command_line_flags}\"\n",
+        "print(cmd)\n",
+        "!$cmd"
```

Comment:
**Issue**: The command construction for launching the web UI is prone to command injection if not properly sanitized. Using Python string formatting with user input without validation can lead to security risks. <br> **Fix**: Ensure that all user inputs are sanitized before being used in command line arguments. Alternatively, use more secure methods to handle command line arguments to prevent potential security vulnerabilities. <br> **Code Suggestion**: 
 ```
 -  cmd += f" --model {output_folder}"
 -  !$cmd
 +  import shlex
 +  safe_output_folder = shlex.quote(output_folder)
 +  cmd += f" --model {safe_output_folder}"
 +  subprocess.run(cmd, shell=True, check=True)
 ```